tasks/base_tasks.py

```diff
@@ -1906,3 +1908,44 @@ def populate(self):
                 )
 
         current_session().execute(stmt)
+
+
+class URLTask(ExternalTask):
+    url = Parameter()
+
+    def output(self):
+        return URLTarget(self.url)
+
+
+class GoogleStorageTask(Task):
+    bucket = Parameter()
+    # The path to the file ("name" in Google Storage conventions)
+    name = Parameter()
+
+    @staticmethod
+    def google_application_credentials():
+        return os.environ.get('GOOGLE_APPLICATION_CREDENTIALS', None)
+
+    def run(self):
+        self._download_blob(self.bucket, self.name, self._destination())
+
+    def _destination(self):
+        return '/bigmetadata/tmp/{}'.format(self.name.split('/')[-1])
+
+    def _download_blob(self, bucket_name, source_blob_name, destination_file_name):
+        """Downloads a blob from the bucket."""
+        storage_client = storage.Client()
+        bucket = storage_client.get_bucket(bucket_name)
+        blob = bucket.blob(source_blob_name)
+
+        LOGGER.info('Downloading blob {} to {}.'.format(
+            source_blob_name,
+            destination_file_name))
+
+        blob.download_to_filename(destination_file_name)
+
+    def complete(self):
+        return os.path.exists(self._destination())
```

Comment:
We should write a proper `exists` method  for `URLTarget` instead of writing a custom `complete` method for the Task

Reply:
This is the comment I was waiting for. I'm not sure about this design, but I couldn't come up with a better one without a deeper refactor. Let me explain how come did I get to a `exists` that just `return True` in `URLTarget`, because I might be wrong but it's intentional.

My problems began at `DownloadGUnzipMC`, which extends `RepoFileGUnzipTask`. That parent class forces you to implement `get_url`, so it implicitly makes an assumption: any potential "repo file" is available with a URL. That's not true for Google Storage: while you can sign files to generate URLs, they only last 7d maximum. The right way to download them is with the credentials and the SDK.

So, I changed the task so instead of assuming the URL that it needed, [it required another URL](https://github.com/CartoDB/bigmetadata/pull/621/commits/1b4319faadc413464c4a830c677e304803f48e49#diff-77ce7c6af6c29a58f760831b1f17072fR164). That URL might be a "hardcoded one" (`from URLTask`, which is just a wrapper for a value), or one provided by `GoogleStorageTask` (which happens to be a "local URL").

That's why `exists` just returns `True`: it's just a value. `URLTask` doesn't check whether the URL actually exists, and `GoogleStorageTask` returns just the value once it's run.


After this explanation, what do you think? Moving `GoogleStorageTask.complete` check for `URLTarget` might make sense, but I'm not so sure about `URLTask`.

Reply:
After a talk, I've implemented this good suggestion :muscle: